### PR TITLE
FBBT Numerics improvements.

### LINKED ISSUE #988: Equal-to-zero tolerance for FBBT?
I am coming across a machine precision issue of having `xl=0` and `xu=-4.440892098500626e-16` as arguments to the following function:

```
def inv(xl, xu):
    if xl <= 0 and xu >= 0:
        return -math.inf, math.inf
    return 1.0/xu, 1.0/xl
```

Should we have a epsilon tolerance for this? @michaelbynum

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:
- This PR primarily involves making FBBT more robust to floating point arithmetic. 
- This PR also creates a PyomoException for other exceptions to inherit from.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/common/errors.py [hunk 1]
```diff
@@ -9,7 +9,11 @@
 #  ___________________________________________________________________________
 
 
-class DeveloperError(NotImplementedError):
+class PyomoException(Exception):
```
Comment (qtothec) [inline pyomo/common/errors.py]: Docstring please?

File: pyomo/contrib/fbbt/fbbt.py [hunk 2]
```diff
@@ -727,6 +733,8 @@ def visiting_potential_leaf(self, node):
                     lb = -math.inf
                 if ub is None:
                     ub = math.inf
+                if lb - self.infeasibility_tol > ub:
+                    raise InfeasibleConstraintException('Variable has a lower bound which is larger than its upper bound: {0}'.format(str(node)))
             self.bnds_dict[node] = (lb, ub)
             return True, None
```
Comment (qtothec) [inline pyomo/contrib/fbbt/fbbt.py]: Out of curiosity, I think `LinearExpression` still presents itself as a leaf node (see https://github.com/Pyomo/pyomo/issues/924). Do we have some kind of handling for that situation?

File: pyomo/contrib/fbbt/fbbt.py [hunk 3]
```diff
@@ -845,10 +871,12 @@ def fbbt_con(con, deactivate_satisfied_constraints=False, integer_tol=1e-5, infe
     if not con.active:
         return ComponentMap()
 
+    interval.feasibility_tol = infeasible_tol
```
Comment (qtothec) [inline pyomo/contrib/fbbt/fbbt.py]: This isn't going to come and bite us later w.r.t. concurrency issues is it?

File: pyomo/contrib/fbbt/fbbt.py [hunk 4]
```diff
@@ -869,7 +897,7 @@ def fbbt_con(con, deactivate_satisfied_constraints=False, integer_tol=1e-5, infe
 
     # check if the constraint is always satisfied
     if deactivate_satisfied_constraints:
-        if lb >= _lb and ub <= _ub:
+        if lb >= _lb - infeasible_tol and ub <= _ub + infeasible_tol:
```
Comment (qtothec) [inline pyomo/contrib/fbbt/fbbt.py]: Doesn't this make the `infeasible_tol` effectively twice the value that it is?

File: pyomo/contrib/fbbt/interval.py [hunk 5]
```diff
@@ -1,11 +1,18 @@
...
+from pyomo.common.errors import DeveloperError, InfeasibleConstraintException, PyomoException
 
 logger = logging.getLogger(__name__)
 
 
+class IntervalException(PyomoException):
+    pass
+
+
+feasibility_tol = 1e-8
```
Comment (qtothec) [inline pyomo/contrib/fbbt/interval.py]: Why `feasibility_tol` here and `infeasibility_tol` in the other module?

File: pyomo/contrib/fbbt/interval.py [hunk 6]
```diff
@@ -19,13 +26,22 @@ def mul(xl, xu, yl, yu):
...
+        ub = math.inf
+    else:
+        ub = 1.0 / xl
+        lb = 1.0 / xu
+    return lb, ub
 
 
 def div(xl, xu, yl, yu):
-    return mul(xl, xu, *inv(yl, yu))
+    if yl <= feasibility_tol and yu >= -feasibility_tol:
```
Comment (qtothec) [inline pyomo/contrib/fbbt/interval.py]: Comment would be nice: "If denominator can include 0, then quotient is unbounded."

File: pyomo/contrib/fbbt/interval.py [hunk 7]
```diff
@@ -19,13 +26,22 @@ def mul(xl, xu, yl, yu):
 
 
 def inv(xl, xu):
-    if xl <= 0 and xu >= 0:
-        return -math.inf, math.inf
-    return 1.0/xu, 1.0/xl
+    if xl <= feasibility_tol and xu >= -feasibility_tol:
```
Comment (qtothec) [inline pyomo/contrib/fbbt/interval.py]: Comment would be nice: "If denominator can include 0, then quotient is unbounded."

**Comment by qtothec** [comment]
I also just opened a PR against this branch which enables basic handling for `_GeneralExpressionData` and `SimpleExpression` nodes.

File: pyomo/contrib/fbbt/fbbt.py (hunk 2 above)
Comment (michaelbynum) [reply pyomo/contrib/fbbt/fbbt.py]: No, ```LinearExpression``` is not currently supported. I will have to add support in another PR.

File: pyomo/contrib/fbbt/fbbt.py (hunk 4 above)
Comment (michaelbynum) [reply pyomo/contrib/fbbt/fbbt.py]: I don't think so. This tests the lower bound of the constraint and the upper bound of the constraint independently. It ensures that, even after the constraint is deactivated, the body of the constraint will not violate the lower bound of the constraint by more than ```feasibility_tol```. Similarly, even after the constraint is deactivated, the body of the constraint will not violate the upper bound of the constraint by more than ```feasibility_tol```.

File: pyomo/contrib/fbbt/interval.py (hunk 5 above)
Comment (michaelbynum) [reply pyomo/contrib/fbbt/interval.py]: fixed

File: pyomo/contrib/fbbt/interval.py (hunk 6 above)
Comment (michaelbynum) [reply pyomo/contrib/fbbt/interval.py]: done

File: pyomo/contrib/fbbt/interval.py (hunk 7 above)
Comment (michaelbynum) [reply pyomo/contrib/fbbt/interval.py]: done

File: pyomo/common/errors.py (hunk 1 above)
Comment (michaelbynum) [reply pyomo/common/errors.py]: Done

**Comment by michaelbynum** [comment]
@qtothec Assuming tests pass, I think this is ready. Sorry for the delay.

File: pyomo/contrib/fbbt/fbbt.py (hunk 2 above)
Comment (carldlaird) [reply pyomo/contrib/fbbt/fbbt.py]: @michaelbynum Have you added an issue about that already?

File: pyomo/contrib/fbbt/fbbt.py [hunk 8]
```diff
@@ -52,11 +55,11 @@
 """
 
 
-class FBBTException(Exception):
+class FBBTException(PyomoException):
     pass
 
 
-def _prop_bnds_leaf_to_root_ProductExpression(node, bnds_dict):
+def _prop_bnds_leaf_to_root_ProductExpression(node, bnds_dict, feasibility_tol):
```
Comment (carldlaird) [inline pyomo/contrib/fbbt/fbbt.py]: @michaelbynum Can we document the keyword argument "feasibility_tol" or remove the documentation for the arguments?
This is true for most (all?) of the private methods that have this change.

File: pyomo/contrib/fbbt/fbbt.py (hunk 2 above)
Comment (michaelbynum) [reply pyomo/contrib/fbbt/fbbt.py]: @carldlaird Yes, there is an existing issue. I just updated it too.

**Comment by michaelbynum** [comment]
@carldlaird Tests are still running, but I believe I have addressed your concerns.

**Pull request merged**